include/jemalloc/internal/sized_alloc_region.h

```diff
@@ -0,0 +1,144 @@
+#ifndef JEMALLOC_INTERNAL_SIZED_ALLOC_REGION_H
+#define JEMALLOC_INTERNAL_SIZED_ALLOC_REGION_H
+
+#include "jemalloc/internal/alloc_ctx.h"
+#include "jemalloc/internal/assert.h"
+#include "jemalloc/internal/atomic.h"
+#include "jemalloc/internal/pages.h"
+
+/*
+ * This file defines sized alloc regions.  These regions have portions of memory
+ * dedicated to a particular size class.  When allocations come out of these
+ * regions, we can quickly determine their size based only on their address and
+ * a little bit of global metadata, thus avoiding any rtree lookups or (if we're
+ * lucky) any other cache-unfriendly lookups at all.
+ *
+ * We orchestrate it so that zero-initialization is a valid state for the
+ * sized_alloc_region_t.  This lets us avoid branching on
+ * opt_sized_alloc_region.
+ */
+
+/*
+ * SIZED_ALLOC_REGION_LG_NUM_SCS:
+ *     The log of the number of size classes that can come out of the
+ *     sized-alloc region.
+ * SIZED_ALLOC_REGION_LG_SC_SIZE:
+ *     The log of the size of virtual address space that we allocate to each
+ *     size class within the sized-alloc region.
+ */
+#if LG_SIZEOF_PTR == 2
```

Comment:
I don't think it's feasible to allow using this with custom hooks, since this is internally at the same "level" -- both the sized alloc region and the extent hooks are just "one source of memory among several". Making this use custom hooks would be hard.

The MAP_HUGETLB thing is tricky, and something we've been talking about a lot (both trying to use hugepages behind the scenes and also to figure out how to work with them in the extent infrastructure -- see #932 and #952). I don't think we have a great story for letting the user enforce a hugepage requirement from a hook in 5.0 (and, we only accidentally have such a story for 4.x).

With respect to limiting the release of memory back to the OS -- have you looked at the decay settings in malloc_conf? You should be able to limit usage pretty effectively by tweaking those settings (or even disabling releasing back to the OS in general, if that's what you'd prefer). If you're implementing an alternate decay policy via the hooks, I'd be curious to hear about it (it could end up as a useful feature for other users, too).